src/Internal/Testnet/DistributeFunds.purs

```diff
@@ -1,252 +1,273 @@
+      { srcWallet = f tx.srcWallet
+      , utxos = map (\utxo -> { wallet: f utxo.wallet, amount: g utxo.amount })
+          tx.utxos
+      }
+
+instance Foldable (Tx wallet) where
+  foldl f a = foldlDefault f a
+  foldr f a = foldrDefault f a
+  foldMap f = foldMap (f <<< _.amount) <<< _.utxos <<< unwrap
+
+instance Traversable (Tx wallet) where
+  sequence = sequenceDefault
+  traverse f (Tx tx) = ado
+    utxos <- traverse
+      (\{ wallet, amount } -> { wallet, amount: _ } <$> f amount)
+      tx.utxos
+    in wrap $ tx { utxos = utxos }
+
+emptyTx :: forall wallet amount. wallet -> Tx wallet amount
+emptyTx srcWallet = wrap
+  { srcWallet
+  , numUtxos: zero
   , utxos: Nil
   }
 
-newtype SourceState src target amount = SourceState
-  { source :: src
+isTxNonEmpty :: forall wallet amount. Tx wallet amount -> Boolean
+isTxNonEmpty (Tx { numUtxos }) = numUtxos > zero
+
+--
+
+type SourceState wallet amount =
+  { srcWallet :: wallet
   , leftover :: amount
-  , tx :: Tx Unit target amount
-  , completeTxs :: List (Tx Unit target amount)
+  , currentTx :: Tx wallet amount
+  }
+
+initSourceState
+  :: forall wallet amount
+   . wallet
+  -> amount
+  -> SourceState wallet amount
+initSourceState srcWallet initFunds =
+  { srcWallet
+  , leftover: initFunds
+  , currentTx: emptyTx srcWallet
   }
 
-derive instance Newtype (SourceState s t a) _
-derive instance Generic (SourceState s t a) _
-derive instance (Eq s, Eq t, Eq a) => Eq (SourceState s t a)
-derive instance (Ord s, Ord t, Ord a) => Ord (SourceState s t a)
-instance (Show s, Show t, Show a) => Show (SourceState s t a) where
+resetSourceTx
+  :: forall wallet amount
+   . SourceState wallet amount
+  -> SourceState wallet amount
+resetSourceTx src = src { currentTx = emptyTx src.srcWallet }
+
+--
+
+data DistrFundsError wallet amount
+  = DistrFunds_MaxUtxosPerTxLowerLimitError
+  | DistrFunds_AssignUtxoError
+      { utxoToAssign :: wallet /\ amount
+      , currentSources :: List (SourceState wallet amount)
+      }
+  | DistrFunds_MaxRoundsExceededError
+
+derive instance Generic (DistrFundsError wallet amount) _
+derive instance (Eq wallet, Eq amount) => Eq (DistrFundsError wallet amount)
+
+instance (Show wallet, Show amount) => Show (DistrFundsError wallet amount) where
   show = genericShow
 
-initialSourceState
-  :: forall src target amount
-   . Semiring amount
-  => { initialFunds :: amount, key :: src }
-  -> SourceState src target amount
-initialSourceState { initialFunds, key } = SourceState
-  { source: key
-  , leftover: initialFunds
-  , tx: emptyTx zero
-  , completeTxs: Nil
+type DistrFundsRoundResult wallet amount =
+  { sources :: List (SourceState wallet amount)
+  , deferredTargets :: List (wallet /\ amount)
   }
 
-parallelizedDistributionPlan
-  :: forall src target amount
-   . Map src (Array (Tx Unit target amount))
-  -> Array (Map src (Tx Unit target amount))
-parallelizedDistributionPlan _ = unsafeThrow "hello"
+data AssignUtxoResult wallet amount
```

Comment:
Please explain the meaning of each alternative

Reply:
Addressed in https://github.com/Plutonomicon/cardano-transaction-lib/pull/1644/commits/b881710f7c5cbda8533a663d9caf5af252fc6329